File: wasm-bench/twosum.c
#include <stddef.h>
#include <stdint.h>

#define new(a, n, t)    (t *)alloc(a, n, sizeof(t), _Alignof(t))
#define assert(c)       while (!(c)) __builtin_trap()

typedef struct {
    char *beg;
    char *off;
    char *end;
} Arena;

[[clang::export_name("alloc")]]
void *alloc(Arena *a, ptrdiff_t count, ptrdiff_t size, ptrdiff_t align)
{
    ptrdiff_t pad = -(uintptr_t)a->off & (align - 1);
    assert(count < (a->end - a->off - pad)/size);
    char *r = a->off + pad;
    a->off += pad + count*size;
    return __builtin_memset(r, 0, count*size);
}

[[clang::export_name("reset")]]
void reset(Arena *a)
{
    a->off = a->beg;
}

[[clang::export_name("newarena")]]
Arena *newarena(ptrdiff_t cap)
{
    size_t old = __builtin_wasm_memory_grow(0, (cap + 0xffffu)>>16);
    assert(old != (size_t)-1);

    char *mem = (char *)(old << 16);
    Arena tmp = {mem, mem, mem+cap};
    Arena *r = new(&tmp, 1, Arena);
    *r = (Arena){tmp.off, tmp.off, tmp.end};
    return r;
}

typedef struct {
    int32_t i;
    int32_t j;
} Solution;

[[clang::export_name("twosum")]]
Solution *twosum(int32_t *nums, int32_t count, int32_t target, Arena *a)
{
    Solution *ret = new(a, 1, Solution);

    int32_t exp = 32 - __builtin_clz(count);

    Arena scratch = *a;
    int32_t *seen = new(&scratch, 1<<exp, int32_t);

    for (int32_t n = 0; n < count; n++) {
        int32_t  number = nums[n];
        int32_t  complement = target - number;
        int32_t  key = complement>number ? complement : number;
        uint32_t hash = key * 489183053u;
        uint32_t mask = (1<<exp) - 1;
        uint32_t step = hash>>(32 - exp) | 1;
        for (int32_t i = hash;;) {
            i = (i + step) & mask;
            int32_t j = seen[i] - 1;
            if (j < 0) {
                seen[i] = n + 1;
                break;
            } else if (nums[j] == complement) {
                ret->i = j;
                ret->j = n;
                return ret;
            }
        }
    }
    return 0;  // no solution
}

File: wasm-bench/twosum.py
# WebAssembly benchmarking
# $ clang --target=wasm32 -O -nostdlib -Wl,--no-entry -o twosum.wasm twosum.c
# $ uv run --with=pywasm3,wasmtime twosum.py
#
# WebAssembly runs much faster than Python, but the context switch is
# relatively expensive, copying data in and out of Wasm memory for each
# operation. The benchmark deliberately pays this cost, with results
# (shared instance only):
#
#             cpython   pypy
#   wasm3:         3x  1.25x
#   wasmtime:     10x  2.00x
#
# That would be typical when replacing hot spots in existing Python with
# equivalent Wasm, leaving the original interface untouched. A 10x boost
# just from that is a good start! Ideally we would change the interface
# to use batching and zero-copy buffer protocol, e.g. np.ndarray backed
# by Wasm memory. In that case typical speed boosts are more in the
# realm of 10x and 100x respectively (2x and 8x for pypy).
#
# That's straightforward single-threaded performance. Running solvers in
# parallel requires separate instances. In the results, "shared" reuses
# one instance between calls, and "fresh" creates a new instance on each
# call. In this benchmark, fresh instantation adds a ~40% overhead for
# wasmtime. So multi-threaded programs should instantate TwoSumWasmtime
# once per thread, then re-use that instance, perhaps caching it in a
# thread-local. This futher complicates Wasm as a transparent, drop-in
# replacement for an existing, thread-safe Python function.
#
# pywasm3 is delivered only as source, and so requires a C compiler on
# the target, which kind of defeats the purpose of doing this in Wasm.
# In contrast, the wasmtime package comes in all flavors of the three
# most popular operating systems and two most popular architectures. So
# you can bundle fast Wasm routines with your pure Python while only
# relying on plain old pip dependencies. It's overall better for this
# particular purpose.

import functools
import random
import struct
import timeit
import wasm3
import wasmtime

class TwoSumPython():
    @staticmethod
    def twosum(nums, target):
        seen = {}
        for i, num in enumerate(nums):
            try:
                return seen[target - num], i
            except:
                seen[num] = i
        return None

class TwoSumWasmtime():
    # NOTE: lazy cache, multi-threaded programs require synchronization
    _module = None

    @classmethod
    def _compile(cls, store):
        if cls._module is None:
            module = wasmtime.Module.from_file(store.engine, "twosum.wasm")
            cls._module = module.serialize()  # save for later
        else:
            module = wasmtime.Module.deserialize(store.engine, cls._module)
        return module

    def __init__(self):
        store    = wasmtime.Store()
        module   = self._compile(store)
        instance = wasmtime.Instance(store, module, ())
        exports  = instance.exports(store)

        newarena     = functools.partial(exports["newarena"], store)
        self._alloc  = functools.partial(exports["alloc"], store)
        self._reset  = functools.partial(exports["reset"], store)
        self._twosum = functools.partial(exports["twosum"], store)
        self._arena  = newarena(1<<30) & 0xffffffff
        self._memory = exports["memory"].get_buffer_ptr(store)

    def twosum(self, nums, target):
        self._reset(self._arena)
        numsptr = self._alloc(self._arena, len(nums), 4, 4) & 0xffffffff
        struct.pack_into(f"<{len(nums)}i", self._memory, numsptr, *nums)
        retptr = self._twosum(numsptr, len(nums), target, self._arena)
        if retptr == 0:
            return None
        return struct.unpack_from("<ii", self._memory, retptr)

class TwoSumWasm3():
    # NOTE: lazy cache, multi-threaded programs require synchronization
    _wasm = None

    @classmethod
    def _parse(cls, env):
        if cls._wasm is None:
            with open("twosum.wasm", "rb") as f:
                cls._wasm = f.read()
        return env.parse_module(cls._wasm)

    def __init__(self):
        env = wasm3.Environment()
        rt  = env.new_runtime(2**10)

        # Loading a module links it with this runtime. It cannot be
        # linked again. To have independant instances, each instance
        # must parse the Wasm image from scratch. That is why the
        # "fresh" bench performs worse than plain Python.
        rt.load(self._parse(env))

        newarena     = rt.find_function("newarena")
        self._alloc  = rt.find_function("alloc")
        self._reset  = rt.find_function("reset")
        self._twosum = rt.find_function("twosum")
        self._arena  = newarena(1<<30) & 0xffffffff
        self._memory = rt.get_memory(0)

    def twosum(self, nums, target):
        self._reset(self._arena)
        numsptr = self._alloc(self._arena, len(nums), 4, 4) & 0xffffffff
        struct.pack_into(f"<{len(nums)}i", self._memory, numsptr, *nums)
        retptr = self._twosum(numsptr, len(nums), target, self._arena)
        if retptr == 0:
            return None
        return struct.unpack_from("<ii", self._memory, retptr)

def bench(name, twosum):
    rng    = random.Random(150)  # chosen for unique solution
    nums   = [rng.randint(-10**9, 10**9) for _ in range(500000)]
    idx    = list(range(len(nums)))
    ai, aj = rng.sample(range(len(nums)), 2)
    target = nums[ai] + nums[aj]

    try:
        with open("nums.txt", "wx") as f:
            print(", ".join(str(n) for n in nums), file=f)
    except:
        pass

    ri, rj = twosum(nums, target)
    assert ri == ai and rj == aj

    time = min(*timeit.repeat(
        stmt="twosum(nums, target)",
        number=10,
        globals={
            "twosum": twosum,
            "nums":   nums,
            "target": target
        },
    ))
    print(f"{name:20}{time:.3g}")

bench("python", TwoSumPython().twosum)
bench("wasmtime-shared", TwoSumWasmtime().twosum)
bench("wasmtime-fresh", lambda n, t: TwoSumWasmtime().twosum(n, t))
bench("wasm3-shared", TwoSumWasm3().twosum)
bench("wasm3-fresh", lambda n, t: TwoSumWasm3().twosum(n, t))
